src/blog/on-removing-let-let-mut.vmd

```diff
@@ -0,0 +1,330 @@
+
+
+We sampled three Vale projects. One had 111 declarations, and only 35 re-assignments. That's only *21%* re-assignments! The other two were even lower, at *20%* and *6%*.
+
+This isn't just Vale either. A randomly chosen Rust library, [Rocket](https://github.com/SergioBenitez/Rocket), had about *8%*. [#rocket]
+
+
+*When did this happen?* We used to re-assign variables all the time!
+
+<slice>
+#rocket: This is approximate; I used `let(\s+mut)?` (4437 results) and `^\s*[\w\[\]\.]+\s*[\+\-\*\/]?=\s+` (351 results) and which may have missed some corners, such as lambdas and mutating call returned values.
+
+Parsing with regular expressions is fun!
+</slice>
+
+
+We suspect three things contributed to this shift over the past couple decades.
+
+
+First, we no longer need re-assignment to return a value from an if-statement. Compare these two snippets in Scala:
+
+<<<<
+```scala
+var weight = 0
+if (i_am_a_potato) {
+  weight = 42
+} else {
+  weight = 73
+}
+```
+////
+```scala
+val weight =
+  if (i_am_a_potato) {
+    42
+  } else {
+    73
+  }
+```
+>>>>
+
+Second, our for-loops became foreach loops, removing that pesky `i++`:
+
+<<<<
+```java
+for (int i = 0; i < ships.size(); i++) {
+  ships[i].launch();
+}
+```
+////
+```java
+for (Spaceship ship : ships) {
+  ship.launch();
+}
+```
+>>>>
+
+Third, we loop over collections less, and now use specialized methods like `find` a lot more. Compare these two snippets in Javascript:
+
+<<<<
+```js
+let foundIndex = -1;
+for (let i in ships) {
+  if (ships[i].name == "Firefly") {
+    foundIndex = i;
+  }
+}
+```
+////
+```js
+let foundIndex =
+    ships.findIndex(
+        x => (x.name == "Firefly"));
+```
+>>>>
+
+
+Either way, we seem to have a lot less re-assignments nowadays, so it makes sense to have the extra keyword on the rarer statement, not the more common one.
+
+
+<slice/>
+
+
+# Another Factor
+
+Note how the declaration doesn't specify whether we can change the variable, like `let` vs `let mut`.
+
+
+One of the benefits of that distinction was that we could easily know whether the variable could change in the future.
+
+
+<<<<
+We actually kept that distinction for a while; we used the `!` symbol, such as `d! = 3;`.
+
+However, we decided to not require it for local variables, because the `set` keyword makes re-assignment more noticeable than it was before.
+
+////
+
+```vale
+func main() {
+  a = 3;
+  b = 3;
+  c = 3;
+  d! = 3;
+  set d = 7;
+  println(d);
+}
+```stdout
+7
+```
+>>>>
+
+
+For example, if we want to know whether the variable `d = 3;` can change, we just need to look for a `set d = 7;` keyword somewhere in the function, which is much more noticeable now than the previous re-assignment syntax was.
+
+
+<<<<
+However, that reasoning doesn't apply to structs. A struct's members might be modified from various far-flung files in our codebase.
+////
+```vale
+struct MyStruct {
+  x! int;
+}
+```
+>>>>
+
+For that reason, we kept the `!` on struct members.
```

Comment:
Maybe add a footnote for other languages that have this already, like OCaml (https://dev.realworldocaml.org/records.html)